### ISSUE TITLE
SQLiteException: too many SQL variables [1.4]

### ISSUE THREAD

**Issue by npomfret** [couchbase/couchbase-lite-android#1091]
I'm getting this error when retrieving docs from a view using the 'keys' parameter.

`Caused by: com.couchbase.lite.internal.database.sqlite.exception.SQLiteException: too many SQL variables (code 1): , while compiling: SELECT key, value, docid, revs.sequence FROM 'maps_40', revs, docs WHERE 1 AND key in (?, ?,...`

I assume it's because there's a limit to the number of keys that can be used in android?  I can't find the docs that say what this limit is.  Any ideas?

**Comment by hideki** [comment]
SQLITE_MAX_VARIABLE_NUMBER is `999`. Did you set more than 999 keys?

```
Maximum Number Of Host Parameters In A Single SQL Statement

A host parameter is a place-holder in an SQL statement that is filled in using one of the sqlite3_bind_XXXX() interfaces. Many SQL programmers are familiar with using a question mark ("?") as a host parameter. SQLite also supports named host parameters prefaced by ":", "$", or "@" and numbered host parameters of the form "?123".

Each host parameter in an SQLite statement is assigned a number. The numbers normally begin with 1 and increase by one with each new parameter. However, when the "?123" form is used, the host parameter number is the number that follows the question mark.

SQLite allocates space to hold all host parameters between 1 and the largest host parameter number used. Hence, an SQL statement that contains a host parameter like ?1000000000 would require gigabytes of storage. This could easily overwhelm the resources of the host machine. To prevent excessive memory allocations, the maximum value of a host parameter number is SQLITE_MAX_VARIABLE_NUMBER, which defaults to 999.

The maximum host parameter number can be lowered at run-time using the sqlite3_limit(db,SQLITE_LIMIT_VARIABLE_NUMBER,size) interface.
```

https://www.sqlite.org/limits.html

**Comment by npomfret** [comment]
Yeah, I was trying 999 but strangely when I maxed my number of keys to 999 it still failed.  I lowered the max a little to 990 and it seems ok.  Don't really understand why, but it's fine for now.  Thanks for your help.

**Comment by hshahdoost** [comment]
By closing this issue, Does it mean it is solved? I think Couchbase should create a safe layer above SQLite which we are allowed to enter as many keys as we want. In my case right now, I need to enter more than 999 keys, But do I have to write my Couchbase logic based on SQLite limitations?

**Comment by hideki** [comment]
@hshahdoost This ticket was closed by the reporter. CBL Andriod 1.x still has this SQLite limitation. If this is the higher priority, we will fix this in next release.

**Comment by lasselaakkonen** [comment]
Coincidentally, I have also run in to this issue just now.

I agree with @hshahdoost that CBL should work around these kinds of limitations for the users of CBL.

@hideki Can this issue be reopened or should I create a new issue?

**Comment by hideki** [comment]
@lasselaakkonen Thank you for your updates. I will reopen this ticket.

**Comment by hideki** [comment]
SQLite has limitation `SQLITE_MAX_VARIABLE_NUMBER`. Avoid this limitation during Java layer.

**Comment by hshahdoost** [comment]
@hideki thanks for reopening the ticket. Though I solved it in my code, I'm waiting to the get the update and remove the exception from my code.
